File: packages/sandbox/daemon/process/log-tee.ts
import {
  closeSync,
  fsyncSync,
  mkdirSync,
  openSync,
  statSync,
  unlinkSync,
  writeSync,
} from "node:fs";
import { dirname } from "node:path";

/**
 * Append-mode tee to a single log file. Each chunk is flushed to the
 * kernel via writeSync; fsync only happens on close to keep per-chunk
 * overhead low. Hard size cap is enforced by **rotation**: when a write
 * would exceed `maxBytes`, the file is unlinked and reopened fresh with a
 * `[log rotated at N bytes]` marker. This keeps per-log disk bounded at
 * roughly `maxBytes` even for long-lived tasks that restart many times
 * (e.g. a dev script crash-looping inside a pod).
 *
 * Open the tee lazily: callers can defer creating the file until the
 * first write so empty logs don't litter the disk.
 */
export class LogTee {
  private fd: number | null = null;
  private written = 0;
  private rotatedOnce = false;

  constructor(
    private readonly path: string,
    private readonly maxBytes: number,
  ) {}

  write(data: string): void {
    if (data.length === 0) return;
    const buf = Buffer.from(data, "utf-8");
    if (this.written + buf.length > this.maxBytes) {
      this.rotate();
    }
    if (!this.openIfNeeded()) return;
    try {
      writeSync(this.fd as number, buf);
      this.written += buf.length;
    } catch {
      /* swallow — log writes must never crash the daemon */
    }
  }

  private rotate(): void {
    this.close();
    try {
      unlinkSync(this.path);
    } catch {
      /* ignore — already gone or never existed */
    }
    this.written = 0;
    this.rotatedOnce = true;
    if (!this.openIfNeeded()) return;
    const marker = Buffer.from(
      `[log rotated at ${this.maxBytes} bytes]\n`,
      "utf-8",
    );
    try {
      writeSync(this.fd as number, marker);
      this.written = marker.length;
    } catch {
      /* ignore */
    }
  }

  writeHeader(label: string): void {
    let prior = this.written;
    if (this.fd === null) {
      try {
        prior = statSync(this.path).size;
      } catch {
        /* new file */
      }
    }
    this.write(
      prior > 0
        ? `\r\n=== ${new Date().toISOString()} ${label} ===\r\n`
        : `${label}\r\n`,
    );
  }

  isTruncated(): boolean {
    return this.rotatedOnce;
  }

  bytesWritten(): number {
    return this.written;
  }

  close(): void {
    if (this.fd === null) return;
    try {
      fsyncSync(this.fd);
    } catch {
      /* fsync best-effort */
    }
    try {
      closeSync(this.fd);
    } catch {
      /* ignore */
    }
    this.fd = null;
  }

  private openIfNeeded(): boolean {
    if (this.fd !== null) return true;
    try {
      mkdirSync(dirname(this.path), { recursive: true, mode: 0o700 });
      // Seed from existing file size so the cap applies across runs that
      // append to the same path (named-script tees rerun without clearing).
      try {
        this.written = statSync(this.path).size;
      } catch {
        /* file doesn't exist yet */
      }
      this.fd = openSync(this.path, "a", 0o600);
      return true;
    } catch {
      return false;
    }
  }
}

File: packages/sandbox/daemon-go/internal/proc/logtee.go
package proc

import (
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// LogTee is an append-mode tee to a single log file. Size is capped by
// rotation-not-truncation: when a write would exceed maxBytes the file is
// unlinked and reopened with a rotation marker.
type LogTee struct {
	mu          sync.Mutex
	path        string
	maxBytes    int64
	f           *os.File
	written     int64
	rotatedOnce bool
}

func NewLogTee(path string, maxBytes int64) *LogTee {
	return &LogTee{path: path, maxBytes: maxBytes}
}

func (t *LogTee) Write(data string) {
	if data == "" {
		return
	}
	t.mu.Lock()
	defer t.mu.Unlock()
	t.writeLocked([]byte(data))
}

func (t *LogTee) writeLocked(buf []byte) {
	if t.written+int64(len(buf)) > t.maxBytes {
		t.rotateLocked()
	}
	if !t.openLocked() {
		return
	}
	if n, err := t.f.Write(buf); err == nil {
		t.written += int64(n)
	}
}

func (t *LogTee) rotateLocked() {
	t.closeLocked()
	os.Remove(t.path)
	t.written = 0
	t.rotatedOnce = true
	if !t.openLocked() {
		return
	}
	marker := fmt.Sprintf("[log rotated at %d bytes]\n", t.maxBytes)
	if n, err := t.f.WriteString(marker); err == nil {
		t.written = int64(n)
	}
}

func (t *LogTee) WriteHeader(label string) {
	t.mu.Lock()
	defer t.mu.Unlock()
	prior := t.written
	if t.f == nil {
		if st, err := os.Stat(t.path); err == nil {
			prior = st.Size()
		}
	}
	if prior > 0 {
		ts := time.Now().UTC().Format("2006-01-02T15:04:05.000Z")
		t.writeLocked([]byte(fmt.Sprintf("\r\n=== %s %s ===\r\n", ts, label)))
	} else {
		t.writeLocked([]byte(label + "\r\n"))
	}
}

func (t *LogTee) IsTruncated() bool {
	t.mu.Lock()
	defer t.mu.Unlock()
	return t.rotatedOnce
}

func (t *LogTee) Close() {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.closeLocked()
}

func (t *LogTee) closeLocked() {
	if t.f == nil {
		return
	}
	t.f.Sync()
	t.f.Close()
	t.f = nil
}

func (t *LogTee) openLocked() bool {
	if t.f != nil {
		return true
	}
	if err := os.MkdirAll(filepath.Dir(t.path), 0o700); err != nil {
		return false
	}
	if st, err := os.Stat(t.path); err == nil {
		t.written = st.Size()
	}
	f, err := os.OpenFile(t.path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o600)
	if err != nil {
		return false
	}
	t.f = f
	return true
}
